[FRAME-5] Generate TypeScript definitions from docs

janvennemann opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/FRAME-5

**Optional Description:**
This is the first version of the TypeScript definition generator. It generates a `index.d.ts` file which can be submitted to https://github.com/DefinitelyTyped/DefinitelyTyped. Currently this needs to be done manually but we could automate this in the future if we are sure the generated typings are stable enough.

janvennemann committed:
[FRAME-5] Generate TypeScript definitions from docs

build commented:
<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.2.0 test /Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9936-NMXLVNHLODS6B2F3BJZSJOJULIBSLLRFPMWKBGWYNTEXIJDLTCNA
> grunt

Running "appcJs:src" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9936-NMXLVNHLODS6B2F3BJZSJOJULIBSLLRFPMWKBGWYNTEXIJDLTCNA/apidoc/lib/typescript_generator.js
   9:2   error    Unnecessarily quoted property 'Dictionary' found  quote-props
  47:10  warning  'hasAllProxyMembers' is defined but never used    no-unused-vars
  58:23  error    'proxyInterface' is not defined                   no-undef
  58:67  error    'proxyInterface' is not defined                   no-undef

/Users/build/jenkins/workspace/sdk_titanium_mobile_PR-9936-NMXLVNHLODS6B2F3BJZSJOJULIBSLLRFPMWKBGWYNTEXIJDLTCNA/android/cli/commands/_buildModule.js
  554:18  warning  Found non-literal argument to RegExp Constructor  security/detect-non-literal-regexp

✖ 5 problems (3 errors, 2 warnings)
  1 error, 0 warnings potentially fixable with the `--fix` option.

Warning: Task "eslint:src" failed. Use --force to continue.

Aborted due to warnings.
npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

hansemannn commented:
@janvennemann Can you fix the ESLint issues? I did locally but also removed the `hasAllProxyMembers` method which I don't want to commit without your permission. Welcome back!

sgtcoolguy commented on apidoc/lib/typescript_generator.js:
```diff
@@ -0,0 +1,542 @@
+const invalidTypeMap = {
+	'2DMatrix': 'Matrix2D',
+	'3DMatrix': 'Matrix3D',
+	'Dictionary': 'any'
```
in our parlance a "Dictionary" is an Object, I believe.

janvennemann replied:
In TypeScript the `object` type only refers to the Object prototype and its properties and methods. Our `Dictionary` refers to any custom object or a specific set of properties/methods of another type, for which i haven't found a proper way of representing that in TypeScript yet. So i just replace it with `any` for now.

janvennemann committed:
[FRAME-5] Remove unused method

janvennemann committed:
[FRAME-5] Allow string values as point coordinates

In various places a point needs to be described in devie specific units or as a percent value for which we need string values.

janvennemann committed:
Remove unnecessary quotes

hansemannn committed:
Merge branch 'master' into FRAME-5

hansemannn commented on apidoc/lib/typescript_generator.js:
```diff
@@ -0,0 +1,528 @@
+ * Map of invalid types and their replacement
+ */
+const invalidTypeMap = {
+	'2DMatrix': 'Matrix2D',
```
We may want to rename the SDK API's for this, either to `Matrix2D` / `Matrix3D` or `Matrix` with `type: Ti.UI.MATRIX_TYPE_2D`. Thoughts?

janvennemann replied:
+1 to rename them. @garymathews  encountered a similar issue when playing around with type constructors instead of factory functions, since any type name starting with a number is invalid.

hansemannn replied:
Which option do you prefer? How do others name it? Any official specification?

janvennemann replied:
From what i have seen, most other libraries only deal with one Matrix at a time, so its simply called Matrix as there is no need to distinguish between 2D and 3D matrices.  Since we need both I'd prefer to name them `Matrix2D` and `Matrix3D` to keep a type for each as they provide a different set of transformation functions you can apply to them.

janvennemann committed:
[FRAME-5] Use method overloads and use inheritance where possible

janvennemann committed:
Merge remote-tracking branch 'origin/FRAME-5' into FRAME-5

janvennemann committed:
[FRAME-5] Use default array type for rest parameters

hansemannn committed:
Merge branch 'master' into FRAME-5

janvennemann committed:
Add more JSDocs

hansemannn committed:
Merge branch 'master' into FRAME-5

janvennemann commented:
@jawa9000 would you mind taking a look at this as it's mostly doc related? Thanks!

hansemannn committed:
Merge branch 'master' into FRAME-5

janvennemann committed:
[FRAME-5] Use correct path for index.d.ts output file

build commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-9936/12/artifact/dist/mobilesdk-7.2.0.v20180419035214-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

hansemannn committed:
Merge branch 'master' into FRAME-5
